### PR TITLE
SF: Adapt SFH_EvaluateRange for multi dataset input

### PR DESCRIPTION
**Pull request by MichaelHuth**
- SFH_EvaluateRange did not accept multi dataset input, which prevented formulas like "epochs(ST) + [1, -2]" to work. This functionality is added.
- Changed SFH_GetSweepsForFormula to accept dataset input only.
- Adapted calling sites.

close https://github.com/AllenInstitute/MIES/issues/1819

- [x] Tests / adapt Tests
- [x] PSX changes
- [x] Documentation
- [x] Handle non-existing (null wave ref) numeric range from `epochs` in SFH_GetSweepsForFormula graciously

### CODE REVIEW

**Comment by t-b** [comment]
I've created the inevitable https://github.com/AllenInstitute/MIES/pull/1989 to add JWN support for wave ref waves.

**Comment by t-b** [comment]
@MichaelHuth 

Why does

```
sel=select(channels(AD9),[31,32,35,37,39],all) 

data(epochs(["E0", "TP"] , $sel), $sel)
```

give `Number of ranges is not equal number of selection.`?

As we support epoch wildcards I would also like to use multiple non-wildcard epochs or?

**Comment by t-b** [comment]
```
# Option 1. Enforce that we only extract epochs 
# from the data selection for matching 
# sweepNo/channelType/channelNumber. 
# Disadvantage: You can't have different 
# selections and breaks documented examples
# 
# Option 2. Introduce offset parmeter
# Disadvantage: every operation needs to get an offset parameter
# data(["E0", "TP"], $sel, offset = [100, 200])
#
# Option 3. Change epochs operation to return
# the number of datasets equal to the number of
# selections with multiple ranges for epochs/
# dataset (our favourite)
data(epochs(["E0", "TP"],$sel) + [[start1, start2],[end1, end2]], $sel)
# 
# Option 4. Don't support it.
```

**Comment by t-b** [comment]
Option 3 it is.

**Comment by t-b** [comment]
- [ ] SFH_Assert in `psxStats` that no two ranges (numeric and epoch) from a sweep intersect

**Comment by t-b** [comment]
Still fixing some test failures.

File: Packages/MIES/MIES_Utilities.ipf [hunk 1]
```diff
@@ -6876,3 +6876,39 @@ Function/WAVE FlattenWaves(WAVE/WAVE input, [variable sdim])
 
 	return output
 End
+
+/// @brief Return the truth if any of the given intervals ]A, B[ intersect.
+///
+/// @param rangesParam Nx2 wave with the interval borders
+threadsafe Function AreIntervalsIntersecting(WAVE rangesParam)
```
Comment (MichaelHuth) [inline Packages/MIES/MIES_Utilities.ipf:6883]: General comment on the function:

I know it is for the ranges here but the function is generalized. So if the function name is about "intervals" the WAVE name should follow that scheme.

Also the for loop seems to be inefficient for large sets of intervals.
How about putting the check in an indexable function and call it through multithread?
The result can then be derived from the sum of the MT return wave.

File: Packages/MIES/MIES_Utilities.ipf [hunk 2]
```diff
@@ -6826,3 +6826,35 @@ Function/WAVE ZapNullRefs(WAVE/WAVE input)
 
 	return result
 End
+
+/// @brief Flatten multidimensional waves inside input to the given dimension
+///
+/// @param input wave reference wave
+/// @param sdim  [optional, defaults to 1] dimensionality to flatten to
+Function/WAVE FlattenWaves(WAVE/WAVE input, [variable sdim])
```
Comment (MichaelHuth) [inline Packages/MIES/MIES_Utilities.ipf:6852]: I think the function name is very generic and should be more specific.

Also with "flatten" I would expect that {{1,2},{3,4} -> {1,2,3,4} (like Redimension/E=1) but it does actually {1,2} and {3,4}.

How about something like "SplitWavesToDimension" ?

**Review by MichaelHuth — COMMENTED** [review]
See comments

File: Packages/MIES/MIES_Constants.ipf [hunk 3]
```diff
@@ -2204,3 +2204,5 @@ StrConstant SF_OP_MERGE   = "merge"
 StrConstant SF_OP_FIT     = "fit"
 StrConstant SF_OP_FITLINE = "fitline"
 StrConstant SF_OP_DATASET = "dataset"
+
+Constant IGOR_MAX_DIMENSIONS = 4
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_Constants.ipf:2208]: Done.

File: Packages/tests/Basic/UTF_Utils.ipf [hunk 4]
```diff
@@ -7522,3 +7522,61 @@ static Function TestGetRowIndex()
...
+		FAIL()
+	catch
+		CHECK_NO_RTE()
+	endtry
+
+	// bails on invalid sdim parameter
+	try
+		Make/FREE wvData = {{1, 2}, {3, 4}}
+		Make/FREE/WAVE wvRef = {wvData}
+		FlattenWaves(wvRef, sdim = 5)
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/tests/Basic/UTF_Utils.ipf:7562]: Done.

File: Packages/MIES/MIES_Utilities.ipf (hunk 2 above)
Comment (t-b) [reply Packages/MIES/MIES_Utilities.ipf:6852]: Done.

File: Packages/MIES/MIES_SweepFormula.ipf [hunk 5]
```diff
@@ -1208,33 +1208,40 @@ static Function [WAVE/WAVE formulaResults, STRUCT SF_PlotMetaData plotMetaData]
...
 
 	variable channelNumber, channelType, sweepNo, isAveraged
 	string channelId, prefix
-	string traceAnnotation
+	string traceAnnotation, annotationPrefix
 
 	prefix = RemoveEnding(ReplaceString(";", plotMetaData.opStack, " "), " ")
 
 	strswitch(plotMetaData.dataType)
+		case SF_DATATYPE_SWEEP:
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_SweepFormula.ipf:1235]: Done. I've also raised an internal ticket for a future linter.

File: Packages/tests/Basic/UTF_SweepFormula.ipf [hunk 6]
```diff
@@ -2693,6 +2693,18 @@ static Function TestOperationData()
...
+	str = "data(epochs(WhatEpochIsThis, " + str + "), " + str + ")"
+	WAVE/WAVE dataWref = SF_ExecuteFormula(str, win, useVariables=0)
+	REQUIRE_EQUAL_VAR(DimSize(dataWref, ROWS), 0)
+
+	// one null range from epochs as TestEpoch1 only exists for sweepNo
+	str = "select(channels(AD),[" + num2istr(sweepNo) + ", " + num2istr(sweepNo + 1) + "],all)"
+	str = "data(epochs([TestEpoch1], " + str + "), " + str + ")"
+	WAVE/WAVE dataWref = SF_ExecuteFormula(str, win, useVariables=0)
+	REQUIRE_EQUAL_VAR(DimSize(dataWref, ROWS), 2)
+
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/tests/Basic/UTF_SweepFormula.ipf:2707]: Fixed.

File: Packages/doc/SweepFormula.rst [hunk 7]
```diff
@@ -1308,16 +1312,17 @@ For example if `select()` is used for the selectData argument then all channels
...
+The output data varies depending on the requested type. Multiple epochs for one
+sweep always result in additional columns.
 
 range:
-Each output data wave is numeric and contains two elements with the start and end time of the epoch in [ms].
+Each output data wave is numeric and has the start/end times in the rows [ms].
 
 name:
-Each output data wave is textual and contains one elements with the name of the epoch.
+Each output data wave is textual and contains name of the epoch.
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/doc/SweepFormula.rst:1322]: I'm not understanding this point. Above I have 

> Multiple epochs for one sweep always result in additional columns.

and then I just describe what is inside that 1xN wave.

File: Packages/MIES/MIES_SweepFormula_Helpers.ipf [hunk 8]
```diff
@@ -1351,16 +1351,16 @@ Function [WAVE adaptedRange, WAVE/T epochRangeNames] SFH_GetNumericRangeFromEpoc
 		return [adaptedRange, $""]
 	endif
 
-	WAVE/T epochNames = range
-	SFH_ASSERT(IsTextWave(epochNames) && !DimSize(epochNames, COLS), "Expected 1d text wave for epoch specification")
+	WAVE/T epochPatterns = range
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_SweepFormula_Helpers.ipf:1354]: Fixed.

File: Packages/MIES/MIES_Utilities.ipf [hunk 9]
```diff
@@ -6876,3 +6876,39 @@ Function/WAVE FlattenWaves(WAVE/WAVE input, [variable sdim])
...
+	endif
+
+	// sort start column in ascending order
+	Duplicate/FREE rangesParam, ranges
+	WaveClear rangesParam
+	SortColumns/KNDX={0} sortWaves=ranges
+
+	for(i = 0; i < numRows; i += 1)
+		ASSERT_TS(IsFinite(ranges[i][0]) && IsFinite(ranges[i][1]), "Expected finite entries")
+		ASSERT_TS(ranges[i][0] < ranges[i][1], "Invalid entry sorting")
```
[the remark this answers is not in the document]
Comment (t-b) [reply Packages/MIES/MIES_Utilities.ipf:6904]: Fixed.

File: Packages/MIES/MIES_Utilities.ipf (hunk 1 above)
Comment (t-b) [reply Packages/MIES/MIES_Utilities.ipf:6883]: I've renamed the function parameter. And also added the MT speedup.

**Review by timjarsky — APPROVED** [review]

File: Packages/doc/SweepFormula.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichaelHuth) [reply Packages/doc/SweepFormula.rst:1322]: ok

**Pull request merged**